helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -428,13 +463,32 @@ public synchronized void setIdealStates(List<IdealState> idealStates) {
   }
 
   /**
-   * Returns the LiveInstances for each of the instances that are curretnly up and running
+   * Returns the LiveInstances for each of the instances that are currently up and running
    * @return
    */
   public Map<String, LiveInstance> getLiveInstances() {
-    return _liveInstanceCache.getPropertyMap();
+    return getLiveInstances(false);
+  }
+
+  /**
+   * Returns the LiveInstances for each of the instances that are currently up and running
+   * @param filterTimedOutInstances - Only set true during maintenance mode. If true, filter out
+   *                                instances that are timed-out during maintenance mode; instances
+   *                                are timed-out if they have been offline for a while before going
+   *                                live during maintenance mode.
+   * @return
+   */
+  public Map<String, LiveInstance> getLiveInstances(boolean filterTimedOutInstances) {
```

Comment:
Can we cache the results somewhere (we may already have it?) instead of computing everytime this method is called?  I think gertLiveInstances are called in many places during one pipeline.

Reply:
Makes sense.